server/app/core/usecases.py

```diff
@@ -144,61 +156,296 @@ async def save(
         context: SaveDatasetContext,
         validator_class: type[IDatasetValidatorService] = DatasetValidatorService,
     ) -> tuple[Dataset, list[str]]:
-        """Create or update a dataset"""
+        """Create or update a dataset — with connector enrichment (in-place update)."""
 
-        # Validate the input dataset
+        # 1 Validate dataset
         validator = validator_class(
             shacl_url=context["shacl_url"],
             ontology_url=context["ontology_url"],
         )
         validator.validate(dataset)
 
-        # Get the local catalog
+        # 2 Get catalog + user
         catalog = await self.repositories.catalogs.get()
-
-        # Get or create the person
         user = context["user"]
         person = await self._get_or_create_person(user)
 
-        # Add metadata from the MMIO file
+        # 3 Add MMIO metadata
         metadata_items, errors = await self._build_mmio_metadata(
             filename, context["oca_uri"]
         )
+
         for metadata in metadata_items:
             dataset += metadata
             dataset.set_attribute(DSPACE.extraMetadata, metadata.uri)
         dataset.set_attribute(DSPACE.metadataFilename, filename)
 
-        # Set the additional attributes for the dataset
+        # 4 Connector integration BEFORE saving
+
+        settings = get_settings()
+        base_url = settings.connector_base_url.rstrip("/")
+        related_folder = context.get("related_data_product")
+
+        # Loop through existing distributions
+        for dist_node in list(dataset.graph.objects(dataset.uri, DCAT.distribution)):
+            access_url = dataset.graph.value(dist_node, DCAT.accessURL)
+            if not access_url:
```

Comment:
if there is no access URL it means the input is not correct, as if he created a distribution it must have access URL or it will cause an issue when it is passed to the training builder